sources/appendix-value-extraction.asciidoc

```diff
@@ -0,0 +1,437 @@
+
+* The container value passed to a value extractor is retrieved from the element annotated with the constraint or `@Valid`:
+
+[source,java]
+----
+private Map<String, @Valid @RetailOrder Order> ordersByName;
+
+public get Map<@NotNull String, Order> getOrdersByName() {
+    return ordersByName;
+}
+----
+
+When validating the `@NotNull` constraint, the map as returned by the getter will be passed to the map key extractor in order to obtain the map keys. When validating the `@RetailOrder` constraint and performing cascaded validation, the map as obtained directly from the field will be passed to the map value extractor in order to obtain the map values.
+
+* Custom extractor for a `Tuple` type:
+
+[source,java]
+----
+public interface Tuple<T1, T2> {
+    T1 getFirst();
+    T2 getSecond();
+}
+----
+
+[source,java]
+----
+public class TupleFirstExtractor implements ValueExtractor<Tuple<@ExtractedValue ?, ?>> {
+
+	@Override
+	public void extractValues(Tuple<?, ?> originalValue, ValueReceiver receiver) {
+		receiver.value( originalValue.getFirst(), "<first>" );
+		}
+	}
+}
+----
+
+[source,java]
+----
+public class TupleSecondExtractor implements ValueExtractor<Tuple<?, @ExtractedValue ?>> {
+
+	@Override
+	public void extractValues(Tuple<?, ?> originalValue, ValueReceiver receiver) {
+		receiver.value( originalValue.getSecond(), "<second>" );
+		}
+	}
+}
+----
+
+[source,java]
+----
+private Tuple<@NotNull @Email String, @NotNull @Min(1) Integer> tuple;
+----
+
+==== Invalid examples
+
+* Extractor types must declare exactly one parameter as the extracted value. Hence this extractor type is invalid:
+
+[source,java]
+----
+public class DoubleExtractor implements ValueExtractor<Multimap<@ExtractedValue ?, @ExtractedValue ?>> { ... }
+----
+
+* No most specific extractor can be found unambiguously:
+
+[source,java]
+----
+public interface CachedValue<V> {
+    V getCachedValue();
+}
+----
+
+[source,java]
+----
+public interface RealValue<V> {
+    V getRealValue();
+}
+----
+
+[source,java]
+----
+public class CachableValue<V> implements CachedValue<V>, RealValue<V> { ... }
+----
+
+[source,java]
+----
+public class CachedValueExtractor implements ValueExtractor<CachedValue<@ExtractedValue ?>> { ... }
+----
+
+[source,java]
+----
+public class RealValueExtractor implements ValueExtractor<RealValue<@ExtractedValue ?>> { ... }
+----
+
+[source,java]
+----
+private CachableValue<@Min(1) Integer> foo;
+----
+
+Validation of `foo` will fail, as none of the two matching extractors is more specific than the other one. An extractor for `CachableValue` must be added, resolving the ambiguity.
+
+==== XML based configuration
+
+TODO
+
+==== Metadata retrieval
+
+TODO
+
+==== Misc.
+
+* Regarding group sequences and default group sequences, the same rules apply for type-level constraints as they apply for regular element-level constraints.
```

Comment:
Add In particular, you can put the @ConvertGroup contraint to a type-use next to a @Valid. Or something approaching.